iothub/device/src/Message.cs

```diff
@@ -2,19 +2,18 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System;
+using System.Collections.Generic;
 using System.IO;
 using System.Threading;
-using Microsoft.Azure.Devices.Client.Extensions;
 using Microsoft.Azure.Devices.Client.Common.Api;
-using System.Collections.Generic;
 using Microsoft.Azure.Devices.Shared;
 
 namespace Microsoft.Azure.Devices.Client
 {
     /// <summary>
     /// The data structure represent the message that is used for interacting with IotHub.
     /// </summary>
-    public sealed class Message : IReadOnlyIndicator, IDisposable
+    public class Message : IReadOnlyIndicator, IDisposable
```

Comment:
One potential issue could be that `SendEventAsync` API accepts `Message, so technically a user could inherit pass in an inherited `Message` instance to that API.
Thoughts?

Reply:
In this instance we have a well defined behavior if a customer sends a TelemetryMessage via SendEventAsync. The GetPayloadObjectBytes() method will throw a nullref since the Convention won't be set.

We can either...

1. Make SendEventAsync() TelemetryMessage aware
2. Make GetPayloadObjectBytes() check the convention before serializing.

Reply:
I think we should add null checks to the convention APIs, to avoid running into unexpected NREs.

I don't think SendEventAsync() needs to be TelemetryMessage aware. My concern is around people inheriting from `Message` and sending their custom implementation to the SendEventAsync() API. I think we can enghance the doc comments to protect against that (add comments to the virtual methods, so that they know what behavior to expect).

Reply:
If you think that is a concern, you can also use composition. TelemetryMessage could have a ```Message message``` class property instead of the inheritance.